src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -53,24 +67,30 @@ public void save(Product product) {
         lock.writeLock().lock();
         try {
             if (product.getId() != null) {
-                productList.set(productList.indexOf(getProduct(product.getId())), product);
+                productList.stream()
+                        .filter(product::equals)
+                        .map(p -> p = product);
```

Comment:
terminal operation is missing here.
"A stream pipeline consists of a source (which might be an array, a collection, a generator function, an I/O channel, etc), zero or more intermediate operations (which transform a stream into another stream, such as filter(Predicate)), and a terminal operation (which produces a result or side-effect, such as count() or forEach(Consumer)). Streams are lazy; computation on the source data is only performed when the terminal operation is initiated, and source elements are consumed only as needed. " [doc](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html)